MidoNet CLI support for license management

alexbikfalvi opened the pull request:
Adds the following object types to the CLI:
- License to show the information for a license bean
- LicenseStatus to show the information for the license
  status

Adds the following commands to the CLI:
- Install to allow the installation of a license object
  from a file
- Status to show the status of an object type, such as
  a license

Includes support for a POST method to the MidoNet API,
where the request media type is binary data (octet stream).

Signed-off-by: Alex Bikfalvi alex.bikfalvi@midokura.com

alexbikfalvi committed:
MidoNet CLI support for license management

Adds the following object types to the CLI:
- License to show the information for a license bean
- LicenseStatus to show the information for the license
  status

Adds the following commands to the CLI:
- Install to allow the installation of a license object
  from a file
- Status to show the status of an object type, such as
  a license

Includes support for a POST method to the MidoNet API,
where the request media type is binary data (octet stream).

Signed-off-by: Alex Bikfalvi <alex.bikfalvi@midokura.com>

alexbikfalvi commented:
The license management CLI specification.
https://docs.google.com/a/midokura.com/document/d/1CHnBA7HAS3MVFKYwmGyASIdmr8EMLseKfByFKMbDUAA/edit#heading=h.6z7wofmdcr5

alexbikfalvi committed:
Update the media type for license management

Signed-off-by: Alex Bikfalvi <alex.bikfalvi@midokura.com>

admytren commented on src/bin/midonet-cli:
```diff
@@ -846,6 +846,8 @@ class Collection(Attr):
                  getter = None,
                  sort_by = None,
                  factory_method = None,
+                 install_method = None,
```
Hmm, can you just re-use "create" or "add" command? This change adds a 3rd verb for "create" operation.

alexbikfalvi replied:
I agree that the adding another "install" command is a bit ugly. Unfortunately, right now the create/add command has a very different syntax, as it works on networks objects, whereas the install command would just receive an object type and a file. I believe the adding the new pattern to the create/add command, would complicate their syntax even more, perhaps making it difficult to understand what each syntax does (on one hand creating objects from scratch, on the other uploading objects from a file).

admytren commented on src/bin/midonet-cli:
```diff
@@ -846,6 +846,8 @@ class Collection(Attr):
                  sort_by = None,
                  factory_method = None,
+                 install_method = None,
+                 status_method = None,
```
Can you tweak "getter" to act like a status command?

alexbikfalvi replied:
I removed the "status" command and used the getter of the LicenseStatus object to allow the user to check the license status via the "show" command. This also changes the syntax of the show command to allow:
show <TYPE>
and therefore the user can use:
show license_status
to check the license status.

admytren commented on src/bin/midonet-cli:
```diff
@@ -2266,6 +2270,54 @@ class VtepBinding(ObjectType):
+        self.put_attr(SingleAttr(name = 'id',
+                                 value_type= UUID(),
+                                 getter = 'get_id'))
+        self.put_attr(SingleAttr(name = 'issuer',
```
Are there other "issuers" besides Midokura?

alexbikfalvi replied:
Right now, I don't think so. Nevertheless, the issuer is a X.500 principal that must be present in a license file, to validate the license. Here, I display the issues, because it is a field present in the license, however this may be redundant.

admytren commented on src/midonetclient/vendor_media_type.py:
```diff
@@ -17,8 +17,9 @@
+# @author: Alex Bikfalvi <alex.bikfalvi@midokura.com>, Midokura
 
-
+APPLICATION_OCTET_STREAM = "application/octet-stream";
```
We typically have a model for most objects, so the license is always a binary blob? Could it be just a uuencoded string in Json object "license=[license string]"?

ryu25ish replied:
@admytren you're right with regards to being consistent with our design, and I would have done what you suggested myself, but this is definitely not wrong as far as the design is concerned, and we also do proper validation in the server side for this blob.  There's really no right answer in this case, but I don't know the original spec to say that which is preferred.  Let's go with this approach for this API.

alexbikfalvi replied:
I like the proposal Artem, and this is something we may change in a future release, perhaps to be able to upload different objects using the same command. Alex from the UI team also suggested to have the binary data encoded to base64.

ryu25ish commented:
@alexbikfalvi thanks for this work!  I think this is pretty much ready to be merged.  Could you answer/address @admytren's comments?  They all seem minor and I'm going to merge this after that.

alexbikfalvi committed:
Removes the Status command for license mgmt

This patch substitues the previous Status command with the
the Show command in order to display the license status. This
changes the syntax of the Show command to allow a pattern
show <TYPE> where type is a CollectionByType, and the command
calls the collection getter without any parameters. With this
patch, the command for checking the license status becomes:

show license_status

This patch also fixes a bug in the Install command, where
the command would always return a License object, regardless
of the collection type used in the command.

Signed-off-by: Alex Bikfalvi <alex.bikfalvi@midokura.com>

ryu25ish commented:
@admytren raised nice points that we should strongly consider.  We can release this as is now but let's keep the discussion going separately.

admytren commented:
Thanks for looking into my comments @ryu25ish and @alexbikfalvi!
